crates/gen-host-js/src/lib.rs

```diff
@@ -418,69 +423,89 @@ impl Js {
+                    return i64ToF64I[0];
                 }
             "),
 
             Intrinsic::Utf8Decoder => self
                 .src
-                .js("const UTF8_DECODER = new TextDecoder('utf-8');\n"),
+                .js("const utf8Decoder = new TextDecoder('utf-8');\n"),
 
-            Intrinsic::Utf8EncodedLen => self.src.js("let UTF8_ENCODED_LEN = 0;\n"),
+            Intrinsic::Utf16Decoder => self
+                .src
+                .js("const utf16Decoder = new TextDecoder(`utf-16${isLE ? '' : 'be'}`);\n"),
+
+            Intrinsic::Utf8EncodedLen => self.src.js("let utf8EncodedLen = 0;\n"),
 
             Intrinsic::Utf8Encode => self.src.js("
-                const UTF8_ENCODER = new TextEncoder('utf-8');
+                const utf8Encoder = new TextEncoder('utf-8');
 
-                function utf8_encode(s, realloc, memory) {
+                function utf8Encode(s, realloc, memory) {
                     if (typeof s !== 'string') \
                         throw new TypeError('expected a string');
 
                     if (s.length === 0) {
-                        UTF8_ENCODED_LEN = 0;
+                        utf8EncodedLen = 0;
                         return 1;
                     }
 
-                    let alloc_len = 0;
+                    let allocLen = 0;
                     let ptr = 0;
                     let writtenTotal = 0;
                     while (s.length > 0) {
-                        ptr = realloc(ptr, alloc_len, 1, alloc_len + s.length);
-                        alloc_len += s.length;
-                        const { read, written } = UTF8_ENCODER.encodeInto(
+                        ptr = realloc(ptr, allocLen, 1, allocLen + s.length);
+                        allocLen += s.length;
+                        const { read, written } = utf8Encoder.encodeInto(
                             s,
-                            new Uint8Array(memory.buffer, ptr + writtenTotal, alloc_len - writtenTotal),
+                            new Uint8Array(memory.buffer, ptr + writtenTotal, allocLen - writtenTotal),
                         );
                         writtenTotal += written;
                         s = s.slice(read);
                     }
-                    if (alloc_len > writtenTotal)
-                        ptr = realloc(ptr, alloc_len, 1, writtenTotal);
-                    UTF8_ENCODED_LEN = writtenTotal;
+                    if (allocLen > writtenTotal)
+                        ptr = realloc(ptr, allocLen, 1, writtenTotal);
+                    utf8EncodedLen = writtenTotal;
+                    return ptr;
+                }
+            "),
+
+            Intrinsic::Utf16Encode => self.src.js("
+                function utf16Encode (str, realloc, memory) {
+                    const len = str.length, ptr = realloc(0, 0, 2, len), out = new Uint16Array(memory.buffer, ptr, len);
+                    let i = 0;
+                    if (isLE) {
```

Comment:
Out of curiosity, is this measurably faster than `TextEncoder` which would handle endianness automatically?

Reply:
I believe `TextEncoder` only supports UTF8.

Reply:
See https://encoding.spec.whatwg.org/#interface-textencoder

Reply:
TIL, makes sense!